File: impl/go/internal/projects/projects.go
// VideoMaker project store + lifecycle + concept-driven director loop (Go).
// Mirrors the TS projects.ts + director.ts: scaffold a Remotion sandbox, turn a
// concept into a complete multi-scene video, launch Studio, code-aware chat.
package projects

import (
	"crypto/rand"
	"encoding/hex"
	"fmt"
	"regexp"
	"strings"
	"sync"
	"time"

	"optimaize-videomaker-go/internal/sandbox"
	"optimaize-videomaker-go/internal/types"
)

type Scene struct {
	ID               string `json:"id"`
	Title            string `json:"title"`
	DurationInFrames int    `json:"durationInFrames"`
	Narration        string `json:"narration"`
	Visual           string `json:"visual"`
	// Per-scene lifecycle so the UI can review each slide live in Studio as soon
	// as its code is written. Each scene is its own Remotion composition, so
	// Studio renders it directly from code (hot-reload) — no per-scene mp4.
	// "ready" = code written and registered in Studio; "error" = compile/repair
	// pending without blocking the other slides.
	Status      string `json:"status,omitempty"`      // pending|writing|ready|error
	RenderError string `json:"render_error,omitempty"`
}
type Storyboard struct {
	Title  string  `json:"title"`
	FPS    int     `json:"fps"`
	Width  int     `json:"width"`
	Height int     `json:"height"`
	Scenes []Scene `json:"scenes"`
}
type Step struct {
	Index     int     `json:"index"`
	Timestamp float64 `json:"timestamp"`
	Phase     string  `json:"phase"`
	Detail    string  `json:"detail"`
	// Optional rich fields so the UI can render a coding-agent transcript.
	Kind     string `json:"kind,omitempty"`     // plan|write_file|command|command_output|repair|info|error
	Path     string `json:"path,omitempty"`     // write_file
	Content  string `json:"content,omitempty"`  // write_file: code written (capped)
	Command  string `json:"command,omitempty"`  // command
	ExitCode *int   `json:"exit_code,omitempty"` // command_output
	Output   string `json:"output,omitempty"`   // command_output (capped)
}

// stepFieldCap bounds large transcript payloads so the polled JSON stays small.
const stepFieldCap = 4000

func capField(s string) string {
	if len(s) <= stepFieldCap {
		return s
	}
	return s[:stepFieldCap] + fmt.Sprintf("\n… (%d more chars)", len(s)-stepFieldCap)
}
type Project struct {
	ID         string            `json:"id"`
	SandboxID  string            `json:"sandbox_id"`
	Prompt     string            `json:"prompt"`
	Reqs       string            `json:"requirements"`
	Goals      string            `json:"goals"`
	State      string            `json:"state"`
	StudioURL  string            `json:"studio_url"`
	Storyboard *Storyboard       `json:"storyboard"`
	ExportPath string            `json:"export_path"`
	Error      string            `json:"error"`
	Steps      []Step            `json:"steps"`
	Chat       []map[string]string `json:"chat"`
	CreatedAt  float64           `json:"created_at"`
	UpdatedAt  float64           `json:"updated_at"`
	mu         sync.Mutex
}

type Store struct {
	mu         sync.Mutex
	projects   map[string]*Project
	mgr        *sandbox.Manager
	studioPort int
}

func NewStore(mgr *sandbox.Manager) *Store {
	return &Store{projects: map[string]*Project{}, mgr: mgr, studioPort: 3100}
}

func now() float64 { return float64(time.Now().UnixNano()) / 1e9 }
func id(p string) string {
	b := make([]byte, 6)
	_, _ = rand.Read(b)
	return p + hex.EncodeToString(b)
}

func (s *Store) step(p *Project, phase, detail string) {
	s.stepRich(p, Step{Phase: phase, Detail: detail})
}

// stepRich appends a step with optional rich transcript fields (capped).
func (s *Store) stepRich(p *Project, st Step) {
	p.mu.Lock()
	st.Index = len(p.Steps)
	st.Timestamp = now()
	st.Content = capField(st.Content)
	st.Output = capField(st.Output)
	p.Steps = append(p.Steps, st)
	p.UpdatedAt = now()
	p.mu.Unlock()
}

func (s *Store) Get(pid string) (*Project, bool) {
	s.mu.Lock()
	defer s.mu.Unlock()
	p, ok := s.projects[pid]
	return p, ok
}
func (s *Store) List() []*Project {
	s.mu.Lock()
	defer s.mu.Unlock()
	out := make([]*Project, 0, len(s.projects))
	for _, p := range s.projects {
		out = append(out, p)
	}
	return out
}
func (s *Store) Manager() *sandbox.Manager { return s.mgr }

func (s *Store) Create(prompt, reqs, goals string) *Project {
	p := &Project{
		ID: id("prj_"), Prompt: prompt, Reqs: reqs, Goals: goals, State: "pending",
		Steps: []Step{}, Chat: []map[string]string{}, CreatedAt: now(), UpdatedAt: now(),
	}
	s.mu.Lock()
	s.projects[p.ID] = p
	s.mu.Unlock()
	go s.scaffold(p)
	return p
}

func (s *Store) scaffold(p *Project) {
	p.State = "scaffolding"
	info, err := s.mgr.Create("process", nil, nil)
	if err != nil {
		p.State = "failed"
		p.Error = err.Error()
		return
	}
	p.SandboxID = info.SandboxID
	cmd1 := "npx --yes create-video@latest --yes --blank ."
	s.stepRich(p, Step{Phase: "scaffold", Detail: "Scaffolding a blank Remotion project", Kind: "command", Command: cmd1})
	r, _ := s.mgr.Backend().Exec(p.SandboxID, cmd1, "", nil, 600)
	ec1 := r.ExitCode
	s.stepRich(p, Step{Phase: "scaffold", Detail: fmt.Sprintf("create-video exited %d", ec1), Kind: "command_output", ExitCode: &ec1, Output: r.Stdout + "\n" + r.Stderr})
	if r.ExitCode != 0 {
		p.State = "failed"
		p.Error = "scaffold failed: " + trim(r.Stderr, 300)
		return
	}
	cmd2 := "npm install --no-audit --no-fund"
	s.stepRich(p, Step{Phase: "install", Detail: "Installing dependencies", Kind: "command", Command: cmd2})
	r, _ = s.mgr.Backend().Exec(p.SandboxID, cmd2, "", nil, 1200)
	ec2 := r.ExitCode
	s.stepRich(p, Step{Phase: "install", Detail: fmt.Sprintf("npm install exited %d", ec2), Kind: "command_output", ExitCode: &ec2, Output: r.Stdout + "\n" + r.Stderr})
	if r.ExitCode != 0 {
		p.State = "failed"
		p.Error = "npm install failed: " + trim(r.Stderr, 300)
		return
	}
	p.State = "ready"
	s.stepRich(p, Step{Phase: "ready", Detail: "Project scaffolded and dependencies installed.", Kind: "info"})
}

func (s *Store) LaunchStudio(pid string) (string, int, error) {
	p, ok := s.Get(pid)
	if !ok {
		return "", 0, fmt.Errorf("project not found")
	}
	if p.StudioURL != "" {
		return p.StudioURL, 0, nil
	}
	s.mu.Lock()
	port := s.studioPort
	s.studioPort++
	s.mu.Unlock()
	s.mgr.Backend().SpawnDaemon(p.SandboxID, fmt.Sprintf("npx --no-install remotion studio --port %d --no-open", port), "", nil)
	p.StudioURL = fmt.Sprintf("http://127.0.0.1:%d", port)
	s.step(p, "studio", fmt.Sprintf("Remotion Studio launched on port %d.", port))
	return p.StudioURL, port, nil
}

// --- file helpers (delegate to the jailed backend) ---
func (s *Store) ListFiles(pid, rel string) ([]types.FileEntry, error) {
	p, ok := s.Get(pid)
	if !ok {
		return nil, fmt.Errorf("project not found")
	}
	return s.mgr.Backend().List(p.SandboxID, rel)
}
func (s *Store) ReadFile(pid, rel string) (string, error) {
	p, _ := s.Get(pid)
	return s.mgr.Backend().ReadFile(p.SandboxID, rel)
}
func (s *Store) WriteFile(pid, rel, content string) error {
	p, _ := s.Get(pid)
	_, err := s.mgr.Backend().WriteFile(p.SandboxID, rel, content)
	return err
}
func (s *Store) RawPath(pid, rel string) (string, error) {
	p, _ := s.Get(pid)
	return s.mgr.Backend().RawPath(p.SandboxID, rel)
}

func trim(s string, n int) string {
	if len(s) <= n {
		return s
	}
	return s[:n]
}

var tsxBlock = regexp.MustCompile("(?s)```(?:tsx|typescript|ts|jsx)?\\s*(.*?)```")

func extractJSON(text string) string {
	// Prefer a fenced block's interior, else the first '{'/'[' to the matching
	// last '}'/']' span. A brace span (not a non-greedy regex) is required so
	// nested objects like the storyboard's scene array survive intact.
	body := text
	if parts := strings.Split(text, "```"); len(parts) >= 3 {
		for i := 1; i < len(parts); i += 2 {
			chunk := strings.TrimSpace(strings.TrimPrefix(strings.TrimSpace(parts[i]), "json"))
			if strings.HasPrefix(chunk, "{") || strings.HasPrefix(chunk, "[") {
				body = chunk
				break
			}
		}
	}
	st := strings.IndexAny(body, "{[")
	if st == -1 {
		return ""
	}
	open := body[st]
	close := byte('}')
	if open == '[' {
		close = ']'
	}
	en := strings.LastIndexByte(body, close)
	if en > st {
		return body[st : en+1]
	}
	return ""
}
func extractTSX(text string) string {
	if m := tsxBlock.FindStringSubmatch(text); m != nil {
		s := strings.TrimSpace(m[1])
		if strings.Contains(s, "export const") || strings.Contains(s, "import") {
			return s
		}
	}
	return strings.TrimSpace(text)
}

// Helpers consumed by director.go (same package).
func (p *Project) setState(st string)        { p.State = st; p.UpdatedAt = now() }
func (s *Store) logStep(p *Project, ph, d string) { s.step(p, ph, d) }

File: impl/ts/src/projects.ts
// VideoMaker project store + lifecycle + concept-driven director loop.
// A "project" is one Remotion sandbox the user edits, previews (Studio), and
// renders. The director loop is the headline: it turns a single concept into a
// complete multi-scene educational video.

import crypto from "node:crypto";
import { getManager, type SandboxManager } from "./sandbox/manager.js";
import { OpenRouterClient } from "./agent/llm-client.js";

export type ProjectState = "pending" | "scaffolding" | "ready" | "generating" | "rendering" | "failed";

export interface ProjectStep {
  index: number;
  timestamp: number;
  phase: string;
  detail: string;
  // Optional rich fields so the UI can render a coding-agent transcript:
  // what the model wrote, what command ran, and its output.
  kind?: "plan" | "write_file" | "command" | "command_output" | "repair" | "info" | "error";
  path?: string; // write_file: the file path
  content?: string; // write_file: the code written (capped)
  command?: string; // command: the shell command
  exitCode?: number; // command_output: process exit code
  output?: string; // command_output: trimmed stdout+stderr (capped)
  frame?: number; // probe: the frame number a per-frame render check targeted
}

// Cap large transcript payloads so the polled project JSON stays bounded.
export const STEP_FIELD_CAP = 4000;
export function capStep(s: string | undefined, n = STEP_FIELD_CAP): string | undefined {
  if (s === undefined) return undefined;
  return s.length > n ? s.slice(0, n) + `\n… (${s.length - n} more chars)` : s;
}

export interface Project {
  id: string;
  sandboxId: string | null;
  prompt: string;
  requirements: string;
  goals: string;
  state: ProjectState;
  studioUrl: string | null;
  studioPort: number | null;
  studioPid: number | null;
  // Version token for the self-written player bundle; bumped on each rebuild so
  // the frontend can bust the iframe cache (?v=<token>).
  studioVersion: number;
  storyboard: Storyboard | null;
  exportPath: string | null;
  error: string | null;
  steps: ProjectStep[];
  chat: Array<{ role: string; content: string }>;
  createdAt: number;
  updatedAt: number;
}

export type SceneStatus = "pending" | "writing" | "ready" | "error";

export interface Scene {
  id: string; // PascalCase component id, e.g. "TitleScene"
  title: string;
  durationInFrames: number;
  narration: string;
  visual: string;
  // Per-scene lifecycle so the UI can review each slide live in Studio as soon
  // as its code is written. Each scene is its own Remotion <Composition>, so
  // Studio renders it directly from code (hot-reload) — no per-scene mp4.
  // "ready" = code written and registered in Studio; "error" = compile/repair
  // pending (set by the repair loop) without blocking the other slides.
  status?: SceneStatus;
  renderError?: string; // trimmed error when status === "error"
}
export interface Storyboard {
  title: string;
  fps: number;
  width: number;
  height: number;
  scenes: Scene[];
}

const STUDIO_PORT_BASE = 3100;

export class ProjectStore {
  private projects = new Map<string, Project>();
  private mgr: SandboxManager;
  private nextStudioPort = STUDIO_PORT_BASE;

  constructor(mgr?: SandboxManager) {
    this.mgr = mgr ?? getManager();
  }

  manager(): SandboxManager {
    return this.mgr;
  }

  get(id: string): Project {
    const p = this.projects.get(id);
    if (!p) throw new Error(`Project not found: ${id}`);
    return p;
  }
  list(): Project[] {
    return [...this.projects.values()].sort((a, b) => b.createdAt - a.createdAt);
  }
  allocStudioPort(): number {
    return this.nextStudioPort++;
  }

  private step(p: Project, phase: string, detail: string, extra: Partial<ProjectStep> = {}): void {
    p.steps.push({
      index: p.steps.length,
      timestamp: Date.now() / 1000,
      phase,
      detail,
      ...extra,
      content: capStep(extra.content),
      output: capStep(extra.output),
    });
    p.updatedAt = Date.now() / 1000;
  }

  // Public transcript logger (by project id) for modules outside this class
  // (studio.ts, probe.ts via the API) to append typed steps.
  logStep(id: string, phase: string, detail: string, extra: Partial<ProjectStep> = {}): void {
    this.step(this.get(id), phase, detail, extra);
  }

  create(prompt: string, requirements: string, goals: string): Project {
    const now = Date.now() / 1000;
    const p: Project = {
      id: `prj_${crypto.randomBytes(6).toString("hex")}`,
      sandboxId: null,
      prompt,
      requirements,
      goals,
      state: "pending",
      studioUrl: null,
      studioPort: null,
      studioPid: null,
      studioVersion: 0,
      storyboard: null,
      exportPath: null,
      error: null,
      steps: [],
      chat: [],
      createdAt: now,
      updatedAt: now,
    };
    this.projects.set(p.id, p);
    void this.scaffold(p.id);
    return p;
  }

  // Scaffold a blank Remotion project + install deps in a fresh sandbox.
  private async scaffold(id: string): Promise<void> {
    const p = this.get(id);
    try {
      p.state = "scaffolding";
      const info = this.mgr.create({ backend: "process" });
      p.sandboxId = info.sandboxId;
      const backend = this.mgr.backendFor(info.sandboxId);

      this.step(p, "scaffold", "Scaffolding a blank Remotion project", { kind: "command", command: "npx create-video@latest --blank ." });
      const scaffold = await backend.exec(info.sandboxId, "npx --yes create-video@latest --yes --blank .", { timeoutS: 600 });
      this.step(p, "scaffold", "create-video finished", { kind: "command_output", exitCode: scaffold.exitCode, output: scaffold.stdout + scaffold.stderr });
      if (scaffold.exitCode !== 0) throw new Error(`scaffold failed: ${scaffold.stderr.slice(0, 300)}`);

      this.step(p, "install", "Installing dependencies", { kind: "command", command: "npm install" });
      const install = await backend.exec(info.sandboxId, "npm install --no-audit --no-fund", { timeoutS: 1200 });
      this.step(p, "install", "npm install finished", { kind: "command_output", exitCode: install.exitCode, output: install.stdout + install.stderr });
      if (install.exitCode !== 0) throw new Error(`npm install failed: ${install.stderr.slice(0, 300)}`);

      p.state = "ready";
      this.step(p, "ready", "Project scaffolded and dependencies installed.");
    } catch (e) {
      p.state = "failed";
      p.error = String((e as Error).message ?? e);
      this.step(p, "error", p.error);
    }
  }

  // Launch Remotion Studio (idempotent) and return the URL.
  launchStudio(id: string): { url: string; port: number } {
    const p = this.get(id);
    if (!p.sandboxId) throw new Error("Project has no sandbox yet.");
    if (p.studioUrl && p.studioPort) return { url: p.studioUrl, port: p.studioPort };
    const port = this.allocStudioPort();
    const backend = this.mgr.backendFor(p.sandboxId);
    const pid = backend.spawnDaemon(
      p.sandboxId,
      `npx --no-install remotion studio --port ${port} --no-open`,
    );
    p.studioPort = port;
    p.studioPid = pid;
    p.studioUrl = `http://127.0.0.1:${port}`;
    this.step(p, "studio", `Remotion Studio launched on port ${port} (pid ${pid}).`);
    return { url: p.studioUrl, port };
  }

  // Files (delegate to the sandbox backend's jailed file ops).
  listFiles(id: string, rel = "."): unknown {
    const p = this.get(id);
    if (!p.sandboxId) throw new Error("No sandbox.");
    return this.mgr.backendFor(p.sandboxId).list(p.sandboxId, rel);
  }
  readFile(id: string, rel: string): string {
    const p = this.get(id);
    if (!p.sandboxId) throw new Error("No sandbox.");
    return this.mgr.backendFor(p.sandboxId).readFile(p.sandboxId, rel, false) as string;
  }
  writeFile(id: string, rel: string, content: string): unknown {
    const p = this.get(id);
    if (!p.sandboxId) throw new Error("No sandbox.");
    return this.mgr.backendFor(p.sandboxId).writeFile(p.sandboxId, rel, content);
  }
  rawPath(id: string, rel: string): string {
    const p = this.get(id);
    if (!p.sandboxId) throw new Error("No sandbox.");
    return this.mgr.backendFor(p.sandboxId).rawPath(p.sandboxId, rel);
  }

  // Collect all source files under src/ for full chat/director context.
  private collectSource(id: string): Record<string, string> {
    const p = this.get(id);
    if (!p.sandboxId) return {};
    const backend = this.mgr.backendFor(p.sandboxId);
    const out: Record<string, string> = {};
    const walk = (rel: string): void => {
      let entries: Array<{ path: string; isDir: boolean }> = [];
      try {
        entries = backend.list(p.sandboxId!, rel) as Array<{ path: string; isDir: boolean }>;
      } catch {
        return;
      }
      for (const e of entries) {
        if (e.isDir) {
          if (!e.path.includes("node_modules")) walk(e.path);
        } else if (/\.(tsx?|css|json)$/.test(e.path) && !e.path.includes("node_modules")) {
          try {
            out[e.path] = backend.readFile(p.sandboxId!, e.path, false) as string;
          } catch {
            /* skip */
          }
        }
      }
    };
    walk("src");
    return out;
  }
}

let store: ProjectStore | null = null;
export function getProjectStore(): ProjectStore {
  if (!store) store = new ProjectStore();
  return store;
}
